### PR TITLE
Feature/require redownload

### PR DESCRIPTION
**Pull request by kasenvr**
Require will now have an option to clear the cache on a dependency when loading it. (effectively cachebusting it, officially...)

This PR is based on #658, so that must be merged first.

### CODE REVIEW

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-cf81af1_d20295c4.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-cf81af1.tar_d20295c4.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-cf81af1.tar_d20295c4.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-cf81af1_d20295c4.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-cf81af1_d20295c4.exe

File: libraries/script-engine/src/ScriptEngine.h [hunk 1]
```diff
@@ -421,10 +422,11 @@ class ScriptEngine : public BaseScriptEngine, public EntitiesScriptEngineProvide
      * @function Script.require
      * @param {string} module - The module to use. May be a JavaScript file, a JSON file, or the name of a system module such 
      *     as <code>"appUi"</code> (i.e., the "appUi.js" system module JavaScript file).
+     * @param {bool} [forceRedownload=false] - Invalidate the cache for this module and redownload it if necessary.
```
Comment (ctrlaltdavid) [inline libraries/script-engine/src/ScriptEngine.h:425]:
```suggestion
     * @param {boolean} [forceRedownload=false] - Invalidate the cache for this module and redownload it if necessary.
```

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-d5eceb8.tar_d25adf9b.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-d5eceb8.tar_d25adf9b.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-d5eceb8_d25adf9b.sh

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-d5eceb8_d25adf9b.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-d5eceb8_d25adf9b.exe

File: libraries/script-engine/src/ScriptEngine.cpp [hunk 2]
```diff
@@ -1875,7 +1876,7 @@ QScriptValue ScriptEngine::require(const QString& moduleId) {
     //   `delete Script.require.cache[Script.require.resolve(moduleId)];`
 
     // cacheMeta is just used right now to tell deleted keys apart from undefined ones
-    bool invalidateCache = module.isUndefined() && cacheMeta.property(moduleId).isValid();
+    bool invalidateCache = (module.isUndefined() && cacheMeta.property(moduleId).isValid()) || forceRedownload;
```
Comment (HifiExperiments) [inline libraries/script-engine/src/ScriptEngine.cpp:1879]:
```suggestion
    bool invalidateCache = forceRedownload || (module.isUndefined() && cacheMeta.property(moduleId).isValid());
```
this way the code can short circuit and save us some work

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-824de17.tar_8a007a4e.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-824de17_8a007a4e.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-824de17.tar_8a007a4e.gz

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-824de17_8a007a4e.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-824de17_8a007a4e.exe

**Review by ctrlaltdavid — APPROVED** [review]

**Comment by vegaslon** [comment]
After talking with implementer of script.require it was suggested that this functionality be implemented as a developer scripting toggle so that it is infinitely more useful and does not require changing java script just to activate and deactivate the cache busting. such as Developer -> Scripting -> Disable Script.require cache. There is already ways to cache bust require before this pr and so if we are going for easy I think that option would be better since it is what web browsers do already for developement.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-7e8dda1.tar_9f3978d3.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-7e8dda1_9f3978d3.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-7e8dda1.tar_9f3978d3.gz

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-7e8dda1_9f3978d3.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-7e8dda1_9f3978d3.exe

File: libraries/script-engine/src/ScriptEngine.cpp [hunk 3]
```diff
@@ -1873,9 +1873,12 @@ QScriptValue ScriptEngine::require(const QString& moduleId) {
     // modules get cached in `Script.require.cache` and (similar to Node.js) users can access it
     // to inspect particular entries and invalidate them by deleting the key:
     //   `delete Script.require.cache[Script.require.resolve(moduleId)];`
+    
+    // Check to see if
```
Comment (ctrlaltdavid) [inline libraries/script-engine/src/ScriptEngine.cpp:1877]: Check to see if ... what?

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]
Also, copyrights need to be added to Menu.cpp, Menu.h, and ScriptEngihe.h.

**Review by ctrlaltdavid — APPROVED** [review]

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-468a20a.tar_05ad3008.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-468a20a_05ad3008.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-468a20a.tar_05ad3008.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-468a20a_05ad3008.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-468a20a_05ad3008.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-4d53c34.tar_05ad3008.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-4d53c34_05ad3008.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-4d53c34.tar_05ad3008.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-4d53c34_05ad3008.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-4d53c34_05ad3008.exe

**Comment by kasenvr** [comment]
Confirm this works over HTTP.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-6450e07.tar_05ad3008.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-6450e07_05ad3008.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-6450e07.tar_05ad3008.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-6450e07_05ad3008.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR664/Vircadia-Alpha-PR664-6450e07_05ad3008.exe

**Comment by kasenvr** [comment]
Confirmed working over HTTP.

**Pull request merged**